server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java

```diff
@@ -254,14 +254,27 @@ private void onFailure(int idx, String nodeId, Throwable t) {
         }
 
         private void finishHim() {
-            threadPool.executor(finalExecutor).execute(() -> newResponse(task, request, responses, listener));
+            if (isCancelled(task)) {
+                listener.onFailure(new TaskCancelledException("task cancelled"));
+                return;
+            }
+
+            final String executor = finalExecutor.equals(ThreadPool.Names.SAME) ? ThreadPool.Names.GENERIC : finalExecutor;
```

Comment:
I'm unsure about this, as we were only dispatching to a `GENERIC` thread if the request was triggered against `0` nodes.

Reply:
The reasons for that seem to go back a long way. I think always forking is ok.